Improve acquisition CTA – pings

vdavid opened the pull request:
Fixes #27100 

Worked with @erzhtor on this one. There's a lot more info under the issue.

erzhtor commented on client/web/src/routes.tsx:
```diff
@@ -8,7 +8,7 @@ import { CodeIntelligenceProps } from './codeintel'
 import { BreadcrumbsProps, BreadcrumbSetters } from './components/Breadcrumbs'
 import type { LayoutProps } from './Layout'
-import type { ExtensionAlertProps } from './repo/RepoContainer'
+import { ExtensionAlertProps } from './repo/actions/InstallIntegrationsAlert'
```
Probably would be better to use `import type ...` as previously. See https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

vdavid replied:
Nice, TIL. :)

erzhtor commented on client/web/src/search/results/StreamingSearchResults.tsx:
```diff
@@ -81,6 +88,25 @@ export const StreamingSearchResults: React.FunctionComponent<StreamingSearchResu
     const patternType = useNavbarQueryState(state => state.searchPatternType)
     const query = useNavbarQueryState(state => state.searchQueryFromURL)
 
+    const [hasDismissedSignupAlert, setHasDismissedSignupAlert] = useLocalStorage(
+        'StreamingSearchResults.hasDismissedSignupAlert',
+        false
+    )
+    const [hasDismissedBrowserExtensionAlert, setHasDismissedBrowserExtensionAlert] = useLocalStorage(
+        'StreamingSearchResults.hasDismissedBrowserExtensionAlert',
+        false
+    )
+    const isBrowserExtensionInstalled = useObservable(browserExtensionInstalled)
+    const displayCTAsBasedOnCadence = usePersistentCadence(CTA_ALERTS_CADENCE_KEY, CTA_ALERT_DISPLAY_CADENCE)
+
+    const onSignupCtaAlertDismissed = useCallback(() => {
+        setHasDismissedSignupAlert(true)
+    }, [setHasDismissedSignupAlert])
+
+    const onBrowserExtensionCtaAlertDismissed = useCallback(() => {
+        setHasDismissedBrowserExtensionAlert(true)
+    }, [setHasDismissedBrowserExtensionAlert])
+
```
Optional, but it would be cool to extract logic into a custom hook to be grouped and easy to understand/read, smth like:
```jsx
const {showSignupAlert, showBrowserExtensionAlert, onAlertDismissed, ...} = useCtaAlert()
```
If this is possible...

vdavid replied:
Extracted this part of the code to a _local_ custom hook. I hope this is what you meant. (It didn't make sense to me to make it more generic than that as this was the only place we're using this functionality.)

erzhtor commented on internal/types/types.go:
```diff
@@ -1035,6 +1035,17 @@ type GrowthStatistics struct {
 }
 
+
+// GrowthStatistics represents the total users that were created,
```
I believe, docs should be updated

erzhtor reviewed (commented):
Good job! 👍  I know it is in the draft state, but I decided to make a 1st early round review and left a few comments.

vdavid commented on client/shared/src/components/CtaAlert.tsx:
```diff
@@ -0,0 +1,42 @@
+import CloseIcon from 'mdi-react/CloseIcon'
+import React from 'react'
+
+import styles from '@sourcegraph/web/src/search/results/StreamingSearchResults.module.scss'
```
Nice, I totally agree, just noticed this after committing the change that fixed this :D

vdavid commented on client/web/src/hooks/usePersistentCadence.test.ts:
```diff
@@ -0,0 +1,46 @@
+    })
+
+    it('returns true on Nth run', () => {
+        localStorage.setItem(testKey, testCadence.toString())
```
Not sure what you mean by "real" local storage.
As far as I know, mocking localStorage is not needed anymore because Jest 24+ mocks it automatically. I didn't find a reliable source for this, but it's mentioned in the `jest-localstorage-mock` package, [here](https://github.com/clarkbw/jest-localstorage-mock#jest-24)
The functionality works right now (the tests pass), so it's definitely saving data _somewhere_. And as Jest is the test runner (there's no browser), I suppose Jest provides the functionality. What would we gain by mocking it explicitly?

vdavid commented on client/web/src/repo/routes.tsx:
```diff
@@ -199,25 +204,32 @@ export const repoRevisionContainerRoutes: readonly RepoRevisionContainerRoute[]
                         defaultBranch={defaultBranch || 'HEAD'}
                         showOnboardingTour={showOnboardingTour}
-                    />
+                    />{' '}
```
It wasn't. Thanks, nice catch!

vdavid commented:
Ready for review!
@rrhyne: [Loom video](https://www.loom.com/share/06f2a491f2744f18865bea49f2dc3465) demoing the feature

More tasks to go:
- [ ] Extract logic to hook as @erzhtor suggested
- [ ] Make sure we're fetching the data from on-prem pings that we need
- [ ] Add example and finalize [RFC](https://docs.google.com/document/d/1RNYu3ZDeT-JGE3rpxfdESUdEKO6ezUKFV5NBhvwd3z8/edit#), then talk to @attfarhan

limitedmage commented on client/shared/src/components/CtaAlert.tsx:
```diff
@@ -0,0 +1,52 @@
+    onClose: () => void
+}
+
+export const CtaAlert: React.FunctionComponent<CtaAlertProps> = props => (
```
Can we add some storybook tests for this new component?

vdavid replied:
Sure, added it, thanks! :)

limitedmage commented on client/web/src/hooks/usePersistentCadence.ts:
```diff
@@ -0,0 +1,21 @@
+export function usePersistentCadence(localStorageKey: string, cadence: number): boolean {
+    if (!incrementedLocalStorageKeys.has(localStorageKey)) {
+        const pageViewCount = parseInt(localStorage.getItem(localStorageKey) || '', 10) || 0
+        localStorage.setItem(localStorageKey, (pageViewCount + 1).toString())
```
Can we use temporary settings instead of localStorage? https://docs.sourcegraph.com/dev/background-information/web/temporary_settings

vdavid replied:
Unfortunately, we cannot. We also use this hook for users who haven't signed in, to set the cadence for our "Hey, come on, sign up!" CTA. The page you linked advises using localStorage in case of unauthenticated users.

limitedmage commented on client/web/src/repo/actions/BrowserExtensionAlert.test.tsx:
```diff
@@ -0,0 +1,12 @@
+    it('matches snapshot', () => {
+        expect(
+            render(<BrowserExtensionAlert className="" onAlertDismissed={() => {}} />).asFragment()
+        ).toMatchSnapshot()
```
I don't think these kinds of snapshot tests without any behavior testing (eg. clicking around to make sure the right things happen) are very useful. I recommend removing this and sticking to storybook tests if no behavior tests are needed.

vdavid replied:
Thanks, I've removed this test. Actually, it existed before the project, I just refactored it to match the new structure, but I'm completely fine with removing it.

limitedmage commented on client/web/src/repo/actions/BrowserExtensionAlert.tsx:
```diff
@@ -0,0 +1,37 @@
+    )
+}
+
+const onBrowserExtensionClick = (): void => {
+    eventLogger.log('InstallBrowserExtensionCTAClicked')
+}
```
Is there a reason why this is not inside the functional component?

vdavid replied:
Yes. If I put it inside the functional component (e.g. to line 18), I'm getting an "ESLint: Move arrow function 'onBrowserExtensionClick' to the outer scope.(unicorn/consistent-function-scoping)" error. Or do you mean something different?

limitedmage commented on client/web/src/repo/actions/InstallIntegrationsAlert.tsx:
```diff
@@ -0,0 +1,71 @@
+}) => {
+    const displayCTABasedOnCadence = usePersistentCadence(CADENCE_KEY, DISPLAY_CADENCE)
+    const isBrowserExtensionInstalled = useObservable<boolean>(browserExtensionInstalled)
+    const [hoverCount] = useLocalStorage<number>(HOVER_COUNT_KEY, 0)
+    const [hasDismissedExtensionAlert, setHasDismissedExtensionAlert] = useLocalStorage<boolean>(
```
These should use temporary settings.

vdavid replied:
We could, but the rest of the data is stored in localStorage.
I think it adds value during debugging/maintenance if we store these two pieces of data with the rest of similar data.
What do you think?

limitedmage commented on client/web/src/repo/actions/NativeIntegrationAlert.test.tsx:
```diff
@@ -0,0 +1,38 @@
+                        }
+                    />
+                ).asFragment()
+            ).toMatchSnapshot()
```
Same comment here re: snapshot tests without behavior. Recommend removing this test and using only storybook.

vdavid replied:
Thanks, I've removed this test too. This also existed before, I just refactored it to match the new structure, but I'm completely fine with removing it.

rrhyne commented:
David, this looks great!

![image](https://user-images.githubusercontent.com/539268/150687735-e843374f-e158-40de-98c7-b5448dafdbda.png)

As to your question about alignment on the image below, if we can just left align the icon with the left align text, that would fix the issue. 

I can't wait to see what impact this has on click-throughs!

vdavid commented:
Thanks for your review @limitedmage! I appreciate all the comments.
I'm usually not the person who pushes back on suggestions, and I feel a bit bad that this is my first code review from you and I couldn't just reply "OK" for all items. There were two suggestions that I couldn't follow and one which I left open for discussion.

vdavid commented:
@attfarhan or @ebrodymoore: I'll need a bit of help here. This feature involved adding **four new fields** in the pings.
@erzhtor and I worked on this. Neither of us had dealt with ping data before, so we tried to do our best by following [the docs](https://docs.sourcegraph.com/dev/background-information/adding_ping_data).

1. Here is [RFC 571](https://docs.google.com/document/d/1RNYu3ZDeT-JGE3rpxfdESUdEKO6ezUKFV5NBhvwd3z8/edit#) that @rrhyne helped put together. The data is aggregated (just four numbers), and it's an easily justified collection of data. I've added the sample. I'm not sure whether I was supposed to copy-paste this JSON data from somewhere, but in the process, I haven't seen any place to test this before writing the RFC, so I just created this sample data manually. If this is right, **I need approval on it**; if not, then some tips on how to fix it up.

2. Here is [a PR for the analytics repo](https://github.com/sourcegraph/analytics/pull/338) where we added the new fields to the BigQuery schema. It's meant to be backward compatible, as per the requirements. **I need approval on it** or tips to test/fix it before

3. This PR is for the ping logic itself. I'll need to test it once the RFC and the analytics PR are approved. I appreciate it if you take a look if it looks generally correct. Hint: **all** go changes are for the ping. The rest is purely front-end.

Next time, I'd do this a bit more step-by-step (RFC → schema → logic), but this was our first, and we just jumped headfirst into the logic. 🤷‍♂️ Any other tips are welcome.

Thanks in advance :)

vdavid commented:
[tl;dr: the RFC and the analytics PR are still ready for review, but this branch is not yet!)]

Update for @attfarhan / @ebrodymoore: A few changes have happened here since my last message:

1. We've separated the front-end changes from the pings. This means that **the review should be more straightforward** because now this branch only contains the ping-related changes, _and_ that it's much **less urgent** now because this doesn't block the front-end release.
2. I've realized I've missed four metrics from the feature. I've added them to the [RFC](https://docs.google.com/document/d/1RNYu3ZDeT-JGE3rpxfdESUdEKO6ezUKFV5NBhvwd3z8/edit#), so now there are a total of **eight** metrics added rather than four. I've also updated the [analytics PR](https://github.com/sourcegraph/analytics/pull/338) with the new metrics. So **these two can still be reviewed**.
3. I'm still working on the SQL in this branch, so this final piece is **not yet ready for review**.

rrhyne reviewed (approved):
Thanks for all your hard work on this. LGTM!

attfarhan commented:
@vdavid sorry I haven't gotten to this. It is on my todo list for today.

attfarhan commented:
@vdavid I have a discussion going on in the RFC about whether we want to collect metrics on CTA clicks, and if so, how long we plan to log this metric. We want to be selective with what we add to pings and I worry these CTAs could be ephemeral (given that CTAs often change), so let me get back to you on this soon.

rrhyne commented:
Farhan, I replied to your comment as to this on the RFC.

attfarhan commented on internal/usagestats/growth.go:
```diff
@@ -105,3 +74,89 @@ FROM
+		return nil, err
+	}
+
+	return &types.CTAMetrics{
```
@vdavid @erzhtor can we nest these so we can potentially add stats for more CTAs in this JSON later? Something like below:
```suggestion
	return &types.CTAMetrics{
	  BrowserExtensionCTA: {
	    DisplayedOnFilePage:
	      UserCount:
	      EventCount:
	    DisplayedOnSearchPage:
	      UserCount:
	      EventCount:
	    ClickedOnFilePage:
	      UserCount:
	      EventCount:
	    ClickedOnSearchPage:
	      UserCount
	      EventCount:
	```

attfarhan replied:
We do something similar for SearchUsage: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/usagestats/aggregated_search.go?L144-152&utm_source=VSCode-1.4.0#L140:6-140:15

attfarhan commented on cmd/frontend/internal/app/updatecheck/handler.go:
```diff
@@ -291,6 +292,7 @@ type pingPayload struct {
 	NewCodeIntelUsage    json.RawMessage `json:"new_code_intel_usage"`
 	SearchUsage          json.RawMessage `json:"search_usage"`
 	GrowthStatistics     json.RawMessage `json:"growth_statistics"`
+	CTAMetrics           json.RawMessage `json:"cta_metrics"`
```
Nit: can we call this `CTAUsage` for consistency?

vdavid commented:
@attfarhan Thanks a lot for your reviews and thoughts! Also, I'm sorry for the late response. This was not a top priority on my agenda, the next release being still some time away.

Responses:
- Nesting, consistent naming (CTAUsage), [ProperCase](https://github.com/sourcegraph/analytics/pull/338#discussion_r794978547): sure, **will do** all these changes!
- I've **responded** in the [RFC thread](https://docs.google.com/document/d/1RNYu3ZDeT-JGE3rpxfdESUdEKO6ezUKFV5NBhvwd3z8/edit#) about cleaning up unused ping data. (I agree on the priorities, and I've brought a solution idea.)

ebrodymoore commented:
All good on my end to merge this. Just updated the schema so these data points can be received!

erzhtor commented:
> All good on my end to merge this. Just updated the schema so these data points can be received!

Hey, @ebrodymoore! Just to clarify the schema change PR is https://github.com/sourcegraph/analytics/pull/338.
P.S: The one you just recently merged https://github.com/sourcegraph/analytics/pull/357#event-6091722788 belongs to this pings implementation https://github.com/sourcegraph/sourcegraph/pull/31379.

attfarhan reviewed (approved):
Ping structure looks good!

ebrodymoore commented:
The schema has been updated (for real, this time)! As long as @erzhtor or @limitedmage is good with this, then it can be merged!
